doc/sphinx/proofs/writing-proofs/proof-mode.rst

```diff
@@ -1,74 +1,176 @@
+The conclusion is a type and each fact has an associated type.
+Facts can be :gdef:`hypotheses <hypothesis>`
+(which are propositions), variables or local definitions.
+The name of the fact appears first.  The type is shown after the colon.  Local
+definitions are shown in the form `n := 0 : nat`, for example, in which `nat` is the
+type of `0`.
 
-   Coq raises this error message when one attempts to use a proof editing command
-   out of the proof editing mode.
+The proof state when you begin proving a theorem shows
+the statement of the theorem below the line and no hypotheses above the
+line:
+
+.. coqtop:: none
+
+   Parameter P: nat -> Prop.
+
+.. coqtop:: out
+
+   Goal forall n m: nat, n > m -> P 1 /\ P 2.
+
+After applying the :tacn:`intros` :term:`tactic`, we see hypotheses above the line.
+The names of variables (`n` and `m`) and hypotheses (`H`) appear before a colon, followed by
+the type they represent.
+
+.. coqtop:: all
+
+   intros.
+
+Some tactics, such as :tacn:`split`, create new goals, which may
+be referred to as :gdef:`subgoals <subgoal>` in descriptions for clarity.
+Goals are numbered from 1 to N at each step of the proof to permit applying a
+tactic to specific goals.
+
+.. coqtop:: all
+
+   split.
+
+"Variables" may refer specifically to facts for which the type of their type
+is `Set` or `Type`, and "hypotheses" to :term:`propositions <proposition>`,
+which are facts for which the type of their type is `Prop` or `SProp`,
+but these terms are also used interchangeably.
+
+.. coqtop:: out
+
+   let t_n := type of n in idtac "type of n :" t_n;
+   let tt_n := type of t_n in idtac "type of" t_n ":" tt_n.
+   let t_H := type of H in idtac "type of H :" t_H;
+   let tt_H := type of t_H in idtac "type of" t_H ":" tt_H.
+
+A proof script, consisting of the tactics that are applied to prove a
+theorem, is often informally referred to as a "proof".
+The real proof, whether complete or incomplete, is a term, the :gdef:`proof term`,
+which users may occasionally want to examine.  (This is based on the
+*Curry-Howard isomorphism* :cite:`How80,Bar81,Gir89,H89`, which is
+a correspondence between between proofs and terms and between
+propositions and types of λ-calculus.  The isomorphism is also
+sometimes called the "propositions-as-types correspondence")
+
+The :cmd:`Show Proof` command displays the incomplete proof term
+before you've completed the proof.  For example:
+
+.. coqtop:: all
+
+   Show Proof.
+
+The incomplete parts, the goals, are represented by
+:term:`existential variables <existential variable>`
+with names that begin with `?Goal`.  The :cmd:`Show Existentials` command
+shows each existential with the hypotheses and conclusion for the associated goal.
+
+.. coqtop:: all
+
+   Show Existentials.
+
+Coq's kernel verifies the correctness of proof terms when it exits
+proof mode by checking that the theorem statement is the same as the
+type of the proof term.
```

Comment:
...by checking that the proof term is well-typed and that its type is the same as the theorem statement.

Reply:
What's the definition of `well-typed`?  Is determining the type of an expression decidable?

Reply:
> What's the definition of `well-typed`?

https://coq.inria.fr/refman/language/core/basic.html#term-well-typed

> Is determining the type of an expression decidable?

Type-inference in Coq is incomplete but that's not the role of the kernel. When a term reaches the kernel, it has sufficient type annotations (e.g. in binders) so that determining its type is decidable and straightforward.

Reply:
Got it.  If Coq tactics can't generate the right binding for a type annotation, can users explicitly provide the binding to work around that?  Is that approach used much?

Reply:
Yes and yes. (Although I'm not used to using the word "binding" in this context: it's confusing because it looks like you are talking about binders.)

Reply:
Maybe I should have said "provide the annotation".  But your earlier comment says "has sufficient type annotations (e.g. in binders)"--that's not talking about binders??

Reply:
Yep yep, sorry for being confusing too.